parcels/spatialhash.py

```diff
@@ -242,69 +305,107 @@ def _barycentric_coordinates(nodes, point, min_area=1e-8):
-    xf_low = xf.min(axis=-1)
-    xf_high = xf.max(axis=-1)
-    xbounds = np.stack([xf_low, xf_high], axis=-1)
+    # Step 4: Third spreading stage
+    # Shift by 4, OR, mask again.
+    # At this point, there are 1 or 2 zeros between many of the bits.
+    n = (n | (n << np.uint32(4))) & np.uint32(0x030C30C3)
+
+    # Step 5: Final spreading stage
+    # Shift by 2, OR, mask.
+    # After this, each original bit is isolated with exactly two zeros
+    # between it and the next bit, ready for 3D Morton interleaving.
+    n = (n | (n << np.uint32(2))) & np.uint32(0x09249249)
 
-    yf = np.stack((lat[:-1, :-1], lat[:-1, 1:], lat[1:, 1:], lat[1:, :-1]), axis=-1)
-    yf_low = yf.min(axis=-1)
-    yf_high = yf.max(axis=-1)
-    ybounds = np.stack([yf_low, yf_high], axis=-1)
+    # Return the dilated value.
+    return n
 
-    return ybounds, xbounds
+
+def _encode_morton3d(x, y, z, xmin, xmax, ymin, ymax, zmin, zmax):
+    """
+    Quantize (x, y, z) to 10 bits each (0..1023), dilate the bits so there are
+    two zeros between successive bits, and interleave them into a 3D Morton code.
+
+    Notes
+    -----
+    - Works with scalars or NumPy arrays (broadcasting applies).
+    - Output is up to 30 bits; we return np.uint32 (or np.uint64 if you prefer).
+    - Requires `part1by2` defined as in your previous snippet.
+    """
+    # Convert inputs to ndarray for consistent dtype/ufunc behavior.
+    x = np.asarray(x)
+    y = np.asarray(y)
+    z = np.asarray(z)
+
+    # --- 1) Normalize each coordinate to [0, 1] over its bounding box. ---
+    # Compute denominators once (avoid division by zero if bounds equal).
+    dx = xmax - xmin
+    dy = ymax - ymin
+    dz = zmax - zmin
+
+    # Normalize to [0,1]; if a range is degenerate, map to 0 to avoid NaN/inf.
+    xn = np.where(dx != 0, (x - xmin) / dx, 0.0)
+    yn = np.where(dy != 0, (y - ymin) / dy, 0.0)
+    zn = np.where(dz != 0, (z - zmin) / dz, 0.0)
+
+    # --- 2) Quantize to 10 bits (0..1023). ---
+    # Multiply by 1023, round down, and clip to be safe against slight overshoot.
+    xq = np.clip((xn * 1023.0).astype(np.uint32), 0, 1023)
+    yq = np.clip((yn * 1023.0).astype(np.uint32), 0, 1023)
+    zq = np.clip((zn * 1023.0).astype(np.uint32), 0, 1023)
+
+    # --- 3) Bit-dilate each 10-bit number so each bit is separated by two zeros. ---
+    # _dilate_bits maps:  b9..b0  ->  b9 0 0 b8 0 0 ... b0 0 0
+    dx3 = _dilate_bits(xq).astype(np.uint64)
+    dy3 = _dilate_bits(yq).astype(np.uint64)
+    dz3 = _dilate_bits(zq).astype(np.uint64)
+
+    # --- 4) Interleave the dilated bits into a single Morton code. ---
+    # Bit layout (from LSB upward): x0,y0,z0, x1,y1,z1, ..., x9,y9,z9
+    # We shift z's bits by 2, y's by 1, x stays at 0, then OR them together.
+    # Cast to a wide type before shifting/OR to be safe when arrays are used.
+    code = (dz3 << 2) | (dy3 << 1) | dx3
+
+    # If you want a compact type, it fits in 30 bits; uint32 is enough.
```

Comment:
Again, who is 'you' here? Rephrase more in line with the rest of Parcels comments?

```suggestion
    # Since our compact type fits in 30 bits, uint32 is enough.
```